src/main/scala/overflowdb/codegen/CodeGen.scala

```diff
@@ -31,63 +24,59 @@ class CodeGen(schemaFile: String, basePackage: String) {
+        case (docMaybe, src) =>
+          val documentation = docMaybe.filter(_.nonEmpty).map(comment => s"""/** $comment */""").getOrElse("")
+          s""" $documentation
+             | $src
+             |""".stripMargin
       }.mkString("\n")
 
       baseDir.createChild(s"$className.java").write(
-        s"""package io.shiftleft.codepropertygraph.generated;
-           |
-           |import overflowdb.*;
+        s"""package $basePackage;
            |
            |public class $className {
            |
-           |$src
+           |$constantsSource
            |}""".stripMargin
       )
     }
 
-    def writeStringConstants(className: String, constants: List[Constant]): Unit = {
-      writeConstantsFile(className, constants) { constant =>
-        s"""public static final String ${constant.name} = "${constant.value}";"""
+    writeConstantsFile("NodeKeyNames", schema.nodeProperties) { property =>
+      (property.comment, s"""public static final String ${property.name} = "${property.name}";""")
+    }
+    writeConstantsFile("EdgeKeyNames", schema.edgeProperties) { property =>
+      (property.comment, s"""public static final String ${property.name} = "${property.name}";""")
+    }
+    writeConstantsFile("NodeTypes", schema.nodeTypes) { nodeType =>
+      (nodeType.comment, s"""public static final String ${nodeType.name} = "${nodeType.name}";""")
+    }
+    writeConstantsFile("EdgeTypes", schema.edgeTypes) { edgeType =>
+      (edgeType.comment, s"""public static final String ${edgeType.name} = "${edgeType.name}";""")
+    }
+    schema.constantsByCategory.foreach { case (category, constants) =>
+      writeConstantsFile(category, constants) { constant =>
+        (constant.comment, s"""public static final String ${constant.name} = "${constant.value}";""")
       }
     }
 
-    def writePropertyKeyConstants(className: String, constants: List[Constant]): Unit = {
-      writeConstantsFile(className, constants) { constant =>
-        val valueType = constant.valueType.getOrElse(throw new AssertionError(s"`valueType` must be defined for Key constant - not the case for $constant"))
-        val cardinality = constant.cardinality.getOrElse(throw new AssertionError(s"`cardinality` must be defined for Key constant - not the case for $constant"))
-        val baseType = valueType match {
-          case "string"  => "String"
-          case "int"     => "Integer"
-          case "boolean" => "Boolean"
+    def propertyKeyCommentAndSource(property: Property): (Option[String], String) = {
+      val src = {
+        val valueType = property.valueType
+        val cardinality = property.cardinality
+        val completeType = cardinality match {
+          case Cardinality.One => valueType
+          case Cardinality.ZeroOrOne => valueType
+          case Cardinality.List => s"scala.collection.Seq<$valueType>"
+          case Cardinality.ISeq => s"immutable.IndexedSeq<$valueType>"
```

Comment:
This pattern matching occurs on multiple different parts with slightly different semantics. Maybe put it as functions into `Cardinality` directly?

Reply:
Hmm, i think i disagree with this one. Pattern matching allows us to keep the Codegen and the Schema uncoupled (for the most part), so I'd prefer to keep it that way. IMO it's also easier to read then, because the code that's being generated is defined all in one place, rather than having to jump to the Schema.